src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/KeyValuePairModelBinder.cs

```diff
@@ -52,37 +63,40 @@ public sealed class KeyValuePairModelBinder<TKey, TValue> : IModelBinder
                 if (bindingContext.IsTopLevelObject)
                 {
                     var model = new KeyValuePair<TKey, TValue>();
-                    return ModelBindingResult.Success(bindingContext.ModelName, model);
+                    bindingContext.Result = ModelBindingResult.Success(bindingContext.ModelName, model);
                 }
 
-                return ModelBindingResult.NoResult;
+                return;
             }
         }
 
         internal async Task<ModelBindingResult> TryBindStrongModel<TModel>(
```

Comment:
This method pushes a new child context because the property name and type are different, so the result is returned and that context is popped when it returns.